### PR TITLE
Vmware common gh

### PR DESCRIPTION
**Pull request by alexbleotu**
### What does this PR do?

Improved connection to a VMware vCenter Server or ESXi host
- added support for passthrough connections via SSPI
- added function to specify if a connection is made to a vCenter or ESXi host
- added function to disconnect from a vCenter or ESXi host
- added tests for changed/new functions
### Tests written?

Yes

### CODE REVIEW

**Comment by alexbleotu** [comment]
the GSSAPI tests are skipped because the gssapi external library is not included. How do I get the library added? (also emailed @cro directly):

```
13:06:09   test_no_gssapi (unit.utils.vmware_test.connection_test.GssapiTokenTest)
13:06:09 [CPU:25.0%|MEM:34.0%|Z:0]  ... SKIP (0.019s)
13:06:09   test_out_token_defined (unit.utils.vmware_test.connection_test.GssapiTokenTest)
13:06:09 [CPU:25.0%|MEM:34.0%|Z:0]  ... SKIP (0.019s)
13:06:09   test_out_token_undefined (unit.utils.vmware_test.connection_test.GssapiTokenTest)
13:06:09 [CPU:25.0%|MEM:34.0%|Z:0]  ... SKIP (0.020s)
13:06:09   test_service_name (unit.utils.vmware_test.connection_test.GssapiTokenTest)
13:06:09 [CPU:33.3%|MEM:34.0%|Z:0]  ... SKIP (0.020s)
```

**Comment by thatch45** [comment]
@rallytime who did the vmware work? That person should review this.
Also, do you think that we need to make explicit vmware exceptions like this?

**Comment by rallytime** [comment]
@thatch45 Most of this work was done by @nmadhok, while I ported his work to the utils file. @gtmanfred has been involved as well (and maybe @cro?). I think a thorough review here from all of us would be worthwhile. I don't think that the explicit VMWare exceptions are needed here, but I am open to hearing why they might be useful.

@alexbleotu In order for those tests to run, we need to get the gssapi lib installed on our jenkins server. (Thanks for writing tests!) @jtand Can you look into that when you get a moment while we wait for reviews?

**Comment by nmadhok** [comment]
I see that a lot of the underlying code has been changed which needs a lot of testing on our end (Clemson University) to make sure things are still working the way they are supposed to. If two different PR's were created (one to add new features and another to change the existing underlying code) then it would have been easier to review it. I'm also curious why the explicit VMware exceptions are needed here and I noticed some changes that were made might only be available on pyVmomi 6.0 and above and might break things for users using pyVmomi < 6.0 so again needs some additional thorough testing on different distro's having different Python versions and different pyVmomi versions. 

I will be travelling to Europe for openSUSE conference and another conference for the next two weeks (until July 5) so might not have the time to review and test this by then. I will still try to test this over the weekend if I find some time.

**Comment by alexbleotu** [comment]
Sorry for not replying earlier - different time zone. The code should be
backwards compatible.

It's the connection code that has changed quite a bit to support
passthrough authentication and also to reconnect if the session expires.
The rest of the changes are small enhancements

Unfortunately, it won't be easy for me to test with old versions of pyVmomi
because we don't have them in our environment.

If I can help with any additional tests on my end, please let me know.

On 17 Jun 2016 05:31, "Nitin Madhok" <notifications@github.com> wrote:

I see that a lot of the underlying code has been changed which needs a lot
of testing on our end (Clemson University) to make sure things are still
working the way they are supposed to. If two different PR's were created
(one to add new features and another to change the existing underlying
code) then it would have been easier to review it. I'm also curious why the
explicit VMware exceptions are needed here and I noticed some changes that
were made might only be available on pyVmomi 6.0 and above and might break
things for users using pyVmomi < 6.0 so again needs some additional
thorough testing on different distro's having different Python versions and
different pyVmomi versions.

I will be travelling to Europe for openSUSE conference and another
conference for the next two weeks (until July 5) so might not have the time
to review and test this by then. I will still try to test this over the
weekend if I find some time.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/saltstack/salt/pull/34059#issuecomment-226678211>, or mute
the thread
<https://github.com/notifications/unsubscribe/AQOKgTGREzdoKrD6X8NzCBZFgE6PthV5ks5qMiMpgaJpZM4I3NHB>
.

**Comment by alexbleotu** [comment]
I am open to taking the VMware specific exceptions out. Why is having them a bad approach? Should I just handle the pyVmomi exceptions where the functions get called (i.e. in the execution module functions)?

**Comment by alexbleotu** [comment]
@nmadhok would it make it easier if I changed this pull request to contain just the connection changes so  there is less to test? I can include the other changes in a different pull request. 

On the pyVmomi version note, even if there is code specific to pyVmomi 6.0, would that be a problem? This goes on develop anyway not on the stable branches and I don't see why the code needs to be kept on older pyVmomi versions on develop. Surely before releasing a new version in production people will test and if the code doesn't work they will update the pyVmomi version.

**Comment by alexbleotu** [comment]
@nmadhok I removed the extra 2 commits - there should be now just the connection change (as well as new functions: a disconnect function and a function that tests whether a connection is made to a vCenter or to a ESXi host).

**Comment by nmadhok** [comment]
@alexbleotu That will definitely make it easier to test. Even though this will be on develop, we still need to support old pyVmomi code and support upto two versions before the latest (6.0.0). I'll test this over the weekend and we should be able to get this merged after the testing. Thanks for the great work!

**Comment by alexbleotu** [comment]
@nmadhok  thanks! I tested  if it connects to vCenters and hosts via username/password and it works fine (I tested when the gssapi lib present and not present). 

I wasn't aware we needed to support 2 previous versions of pyVmomi. As I said before, we don't have the previous versions in our environment so it's not easy for me to test. If you could test that, that would be great.

**Comment by cachedout** [comment]
@nmadhok Just checking to see if you got to this over the weekend. Thanks!

**Comment by nmadhok** [comment]
Not yet! Was busy preparing my presentation for openSUSE conference in Germany. Gong to get on this as soon as possible. Sorry for the delay! 

--

Thanks,
Nitin Madhok
Clemson University

CONFIDENTIALITY NOTICE
This e-mail, and any attachments thereto, is intended only for use by the addressee(s) named herein and may contain privileged and/or confidential information. If you are not the intended recipient of this e-mail, any dissemination, distribution or copying of this e-mail, and any attachments thereto, is strictly prohibited. If you have received this e-mail in error, please immediately notify the sender by e-mail or telephone and permanently delete all copies of this e-mail and any attachments.

**Comment by cachedout** [comment]
@nmadhok Just checking. Say hi to @thatch45 when you're there and have fun!

**Comment by cachedout** [comment]
Hi @nmadhok Just checking with you here. Thanks!

**Comment by nmadhok** [comment]
Still in Europe. Get back to work on July 5. I'll test this as soon as I'm back. Sorry for delaying this.

--

Thanks,
Nitin Madhok
Clemson University

CONFIDENTIALITY NOTICE
This e-mail, and any attachments thereto, is intended only for use by the addressee(s) named herein and may contain privileged and/or confidential information. If you are not the intended recipient of this e-mail, any dissemination, distribution or copying of this e-mail, and any attachments thereto, is strictly prohibited. If you have received this e-mail in error, please immediately notify the sender by e-mail or telephone and permanently delete all copies of this e-mail and any attachments.

**Comment by cachedout** [comment]
@nmadhok Thanks!

@alexbleotu Could you please rebase this? It now has merge conflicts. Thanks.

**Comment by alexbleotu** [comment]
Hey guys, I've been on vacation. I will rebase today

**Comment by cachedout** [comment]
@alexbleotu Great! Many thanks.

**Comment by alexbleotu** [comment]
My tests fail after applying the change done by @cro (https://github.com/saltstack/salt/commit/14ea29f446563c1c8db255e0a458eaaeeaac5c67) - funny enough this caused the merge conflicts, too. 

I will wait for @cro to comment

File: salt/utils/vmware.py [hunk 1]
```diff
@@ -79,7 +80,7 @@
 import time
 
 # Import Salt Libs
-from salt.exceptions import SaltSystemExit
+import salt.exceptions as excs
```
Comment (nmadhok) [inline salt/utils/vmware.py]: This should be changed to the following I think so it easier to understand the namespace wherever it gets referenced:
```
import salt.exceptions
```

File: salt/utils/vmware.py [hunk 2]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
-            return service_instance
-        Disconnect(service_instance)
-
+def _get_service_instance(host, username, password, protocol,
+                          port, mechanism, principal, domain):
+    '''
+    Internal method to authenticate with a vCenter server or ESX/ESXi host
+    and return the service instance object.
+    '''
+    log.debug('Retrieving new service instance')
```
Comment (nmadhok) [inline salt/utils/vmware.py]: This should go to the trace log instead of the debug log.

File: salt/utils/vmware.py [hunk 3]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+    token = None
+    if mechanism == 'userpass':
```
Comment (nmadhok) [inline salt/utils/vmware.py]: userpass mechanism should be the default if no mechanism is specified instead of raising an exception.

File: salt/utils/vmware.py [hunk 4]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+            raise excs.CommandExecutionError('Login mechanism userpass was '
+                                             'specified but the mandatory '
+                                             'parameter \'password\' is '
+                                             'missing')
+
+        if domain and '@' not in username and '\\' not in username:
+            # Append the domain if the username doesn't already contain a
+            # domain (either old or new style representation)
+            username = '{0}@{1}'.format(username, domain)
+            log.debug('Domain added to username: \'{0}\''.format(username))
```
Comment (nmadhok) [inline salt/utils/vmware.py]: No need to log this.

File: salt/utils/vmware.py [hunk 5]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+                raise excs.VMwareConnectionError(str(exc))
+        else:
+            err_msg = 'Login mechanism \'{0}\' was specified but the' \
+                      ' mandatory parameters are missing'.format(mechanism)
+            raise excs.CommandExecutionError(err_msg)
+    else:
+        raise excs.CommandExecutionError('Unsupported mechanism: \'{0}\''
+                                         ''.format(mechanism))
     try:
+        log.debug('Connecting using the \'{0}\' mechanism, with username '
```
Comment (nmadhok) [inline salt/utils/vmware.py]: This needs to go to trace log instead of debug log.

File: salt/utils/vmware.py [hunk 6]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
-                )
-                ssl._create_default_https_context = default_context
-            elif (isinstance(exc, vim.fault.HostConnectFault) and 'SSL3_GET_SERVER_CERTIFICATE\', \'certificate verify failed' in exc.msg) or 'SSL3_GET_SERVER_CERTIFICATE\', \'certificate verify failed' in str(exc):
+            if (isinstance(exc, vim.fault.HostConnectFault) and
+                '[SSL: CERTIFICATE_VERIFY_FAILED]' in exc.msg) or \
+               '[SSL: CERTIFICATE_VERIFY_FAILED]' in str(exc):
+
                 import ssl
-                default_context = ssl._create_default_https_context
-                ssl._create_default_https_context = ssl._create_unverified_context
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Have you tested this specific case? You're removing these lines and passing the unverified context directly to the SmartConnect method. As far as I remember, there was a reason why it was being done in this way and not being passed directly. Want to make sure that you have tested this specific case and that it has the same behavior.

File: salt/utils/vmware.py [hunk 7]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+    '''
+    Authenticate with a vCenter server or ESX/ESXi host and return the service instance object.
+
+    host
+        The location of the vCenter server or ESX/ESXi host.
+
+    username
+        The username used to login to the vCenter server or ESX/ESXi host.
+        Required if mechanism is ``userpass``
+    password
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Enter a new line before this line

File: salt/utils/vmware.py [hunk 8]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+        The password used to login to the vCenter server or ESX/ESXi host. This
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Remove `` This`` from the end of this line since it doesn't make sense

File: salt/utils/vmware.py [hunk 9]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+        Required if mechanism is ``sspi``
```
Comment (nmadhok) [inline salt/utils/vmware.py]: This should be changed to:
```
Required if mechanism is ``userpass``
```

File: salt/utils/vmware.py [hunk 10]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+    protocol
+        Optionally set to alternate protocol if the vCenter server or ESX/ESXi host is not
+        using the default protocol. Default protocol is ``https``.
+
+    port
+        Optionally set to alternate port if the vCenter server or ESX/ESXi host is not
+        using the default port. Default port is ``443``.
+
+    mechanism
+        pyVmomi connection mechanism. Can either be ``userpass`` or ``sspi``
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Also add the following:
```
Default mechanism is ``userpass``
```

File: salt/utils/vmware.py [hunk 11]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+                                             'specified but the mandatory '
+                                             'parameter \'username\' is '
+                                             'missing')
+        if password is None:
+            raise excs.CommandExecutionError('Login mechanism userpass was '
+                                             'specified but the mandatory '
+                                             'parameter \'password\' is '
+                                             'missing')
+
+        if domain and '@' not in username and '\\' not in username:
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Why is this needed? Having multiple ways of specifying username in the cloud provider configuration would cause confusion. Also, specifying domain is not needed for ``userpass`` mechanism.

File: salt/utils/vmware.py [hunk 12]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+                                                 protocol,
+                                                 port,
+                                                 mechanism,
+                                                 principal,
+                                                 domain)
+
+    return service_instance
+
+
+def disconnect(service_instance):
```
Comment (nmadhok) [inline salt/utils/vmware.py]: No need for this function. ``Disconnect`` is already imported into the namespace.

File: salt/utils/vmware.py [hunk 13]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+    service_instance
+        The Service Instance from which to obtain managed object references.
+    '''
+
+    Disconnect(service_instance)
+
+
+def is_connection_to_a_vcenter(service_instance):
+    '''
+    Function that returns True if a connection is made to a vCenter Server and
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Change to:
```
Function that returns True if the connection is made to a vCenter Server and
```

File: salt/utils/vmware.py [hunk 14]
```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+    False if the connection iss made to an ESXi host
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Change spelling of ``iss`` to ``is``

File: salt/utils/vmware.py (hunk 6 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: I tested connections both with unverified and verified contexts, but with pyVmomi 6.0. I can't test with older versions unfortunately

File: salt/utils/vmware.py (hunk 1 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: done

File: salt/utils/vmware.py (hunk 2 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: done

File: salt/utils/vmware.py (hunk 4 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: removed if clause

File: salt/utils/vmware.py (hunk 7 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: done

File: salt/utils/vmware.py (hunk 8 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: Done

File: salt/utils/vmware.py (hunk 9 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: Done

File: salt/utils/vmware.py (hunk 10 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: Done

File: salt/utils/vmware.py (hunk 11 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: You are right. Removed.

File: salt/utils/vmware.py (hunk 12 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: You are right. Removed.

File: salt/utils/vmware.py (hunk 13 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: done

File: salt/utils/vmware.py (hunk 14 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: Done

**Comment by alexbleotu** [comment]
@nmadhok  Thanks for the review! I have answered all items/fixed all issues. 
@cro I still have the lines you removed in the code as it still breaks for me when I test. Did you have a chance to look into this?

File: salt/utils/vmware.py (hunk 5 above)
Comment (alexbleotu) [inline salt/utils/vmware.py]: done

File: salt/utils/vmware.py [hunk 15]
```diff
@@ -158,7 +166,105 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
...
+    if mechanism == 'userpass':
+        if username is None:
+            raise salt.exceptions.CommandExecutionError(
+                'Login mechanism userpass was specified but the mandatory '
+                'parameter \'username\' is missing')
+        if password is None:
+            raise salt.exceptions.CommandExecutionError(
+                'Login mechanism userpass was specified but the mandatory '
+                'parameter \'password\' is missing')
+
```
Comment (nmadhok) [inline salt/utils/vmware.py]: Get rid of this newline too

Comment (alexbleotu) [inline salt/utils/vmware.py]: done

**Comment by nmadhok** [comment]
@cachedout Everything looks good to me now. We're just waiting for @cro to respond to @alexbleotu

**Comment by cachedout** [comment]
@nmadhok Thanks!

@cro Could you please have a look?

**Comment by cro** [comment]
Sure thing.

**Comment by alexbleotu** [comment]
Can we run the tests again? The failed tests aren't specific to my change. I believe they were because the python version they ran with was < 2.7.9 (where the SSLContext class was added). Unfortunately the results are gone, so this is from memory.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by alexbleotu** [comment]
Same issue:
`23:15:43          Python: 2.7.6 (default, Jun 22 2015, 17:58:13)` (from the ubuntu test suite)

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
Go Go Jenkins!

File: salt/utils/vmware.py [hunk 16]
```diff
@@ -187,71 +303,61 @@ def get_service_instance(host, username, password, protocol=None, port=None):
     service_instance = GetSi()
     if service_instance:
         if service_instance._GetStub().host == ':'.join([host, str(port)]):
-            return service_instance
+            log.trace('Using cached service instance')
+            if salt.utils.is_proxy():
```
Comment (alexbleotu) [inline salt/utils/vmware.py]: Removed as part of 14ea29f446563c1c8db255e0a458eaaeeaac5c67, but not sure it works

Comment (alexbleotu) [inline salt/utils/vmware.py]: @cro This is the change that made the connection code fail in my tests.

**Comment by alexbleotu** [comment]
Hey guys, any updates on this?

**Comment by nmadhok** [comment]
@cro ping!

**Comment by cachedout** [comment]
@cro I would really like to get this in. Could you please respond to @nmadhok ? Thanks.

**Comment by cro** [comment]
Go Go Jenkins!

File: salt/utils/vmware.py (hunk 16 above)
Comment (cro) [inline salt/utils/vmware.py]: I see why now, it took a little while to track that down.  I think we can merge this.

**Comment by cro** [comment]
Go Go Jenkins!

**Comment by alexbleotu** [comment]
thanks @cro 
could we run the ubuntu test suite on 2.7.9, so the tests past?

**Comment by cro** [comment]
I'm sorry @alexbleotu -- I don't understand.  Do you mean run the automated suite on 2.7.9 because 2.7.9 and following has the backported SSL components from later Pythons?

So this PR will not function on Py 2.6 or anything earlier than Py 2.7.9?

**Comment by alexbleotu** [comment]
@cro Yeah, the code does not work with python earlier than 2.7.9. It requires an ssl function only present from 2.7.9. This was not introduced in this PR, it has always been there, but it's highlighted now by the unit tests I added.

**Comment by nmadhok** [comment]
@alexbleotu Well that is an issue. The tests or any code needs to be backwards compatible with Python 2.6 since that is the default that get's shipped with a lot of distributions that are in use in production.

**Comment by nmadhok** [comment]
@alexbleotu @cro I just noticed that this PR changes the way SSL Exceptions are handled. This would break things for Python < 2.7.9 and shouldn't be merged until that is fixed.

**Comment by alexbleotu** [comment]
@nmadhok  @cro 

Ok, I wasn't aware of this requirement. I will try to fix it.

I am not convinced it currently works with Python 2.6 tho .

**Comment by cro** [comment]
I've been looking into this today.  It may be possible, not sure yet.

**Comment by cro** [comment]
Part of the challenge is the fact that VMware left behind Py 2.6 compat entirely with PyVmomi 6+.

**Comment by alexbleotu** [comment]
I will try and fix it on Monday and update the tests in the pull request.

**Comment by cro** [comment]
@alexbleotu Thank you--I did take a brief crack at this and attempted to get it working with the backports.ssl package.  I didn't make a lot of progress.  Looking forward to seeing if you have better luck than I did. :-)

**Comment by alexbleotu** [comment]
@cro, @nmadhok  I checked the connection to both a vcenter with SSL certs and one without using python 2.7.3. With any python < 2.7.9 the connection doesn't fail at all because SSL cert validation is disabled by default. This was introduced in 2.7.9 (https://www.python.org/dev/peps/pep-0476/). Thus the connection simply works for me without raising any exception.  

The exceptions are only triggered in 2.7.9 where the validation is enabled, and the exception handling covers those exceptions. I can add an additional condition to check for the python version, but am not sure in which of your tests the connection fails when using python < 2.7.9.

**Comment by alexbleotu** [comment]
Fixed python > 2.7.9 specific tests to be skipped if the suite is run with a previous python. The rest of the tests that fail are not to do with this pull request.

**Comment by cro** [comment]
Thanks @alexbleotu -- nice work.
